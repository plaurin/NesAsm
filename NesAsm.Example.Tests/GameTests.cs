using ApprovalTests.Reporters;
using NesAsm.Analyzers;
using NesAsm.Analyzers.Tests;
using System.Diagnostics;
using Xunit;
using Xunit.Abstractions;

namespace NesAsm.Example.Tests;

[UseReporter(typeof(VisualStudioReporter))]
public class GameTests
{
    private readonly ITestOutputHelper _output;

    public GameTests(ITestOutputHelper output)
    {
        _output = output;
    }

    [Fact]
    public void TestHiloWorldGame()
    {
        TestTranspileGame(@"HiloWorld\HiloWorld.cs");

        var gameInfo = new GameInfo { GameFolder = "HiloWorld", AsmFileToCompile = "HiloWorld.s", NesOutputFile = "hiloworld.nes" };
        TestCompileGame(gameInfo);
    }

    [Fact]
    public void TestGame1()
    {
        TestTranspileGame(@"Game1\Game1.cs", "Controller.cs");

        var gameInfo = new GameInfo { GameFolder = "Game1", AsmFileToCompile = "game1.s", NesOutputFile = "game1.nes" };
        TestCompileGame(gameInfo);
    }

    [Fact]
    public void TestGame1C()
    {
        TestTranspileGame(@"Game1\Game1C.cs", @"Game1\Game1Char.cs", "Controller.cs", "PPU.cs");

        var gameInfo = new GameInfo { GameFolder = "Game1", AsmFileToCompile = "game1c.s", NesOutputFile = "game1c.nes" };
        TestCompileGame(gameInfo);
    }

    [Fact]
    public void TestGame2C()
    {
        TestTranspileGame(@"Game2\Game2C.cs", @"Game2\Game2Char.cs", "Controller.cs", "PPU.cs");

        var gameInfo = new GameInfo { GameFolder = "Game2", AsmFileToCompile = "game2c.s", NesOutputFile = "game2c.nes" };
        TestCompileGame(gameInfo);
    }

    private static void TestTranspileGame(params string[] filenames)
    {
        var asmGenerator = new AsmGenerator();
        var charGenerator = new CharGenerator();

        var sourcePaths = filenames.Select(filename => Path.Combine(@"..\..\..\..\NesAsm.Example", filename)).ToArray();

        GeneratorTestUtilities.TestGenerator(sourcePaths, asmGenerator, charGenerator);
    }

    private void TestCompileGame(GameInfo gameInfo)
    {
        var psi = new ProcessStartInfo()
        {
            WorkingDirectory = Path.GetFullPath($@"..\..\..\..\NesAsm.Example\{gameInfo.GameFolder}"),
            FileName = @"c:\cc65\bin\cl65",
            Arguments = $"--verbose --target nes -o {gameInfo.NesOutputFile} {gameInfo.AsmFileToCompile}",
            RedirectStandardOutput = true,
            RedirectStandardError = true,
        };

        var process = Process.Start(psi);

        process!.OutputDataReceived += (s, e) => _output.WriteLine(e.Data ?? "");
        process.ErrorDataReceived += (s, e) => _output.WriteLine(e.Data != null ? $"ERR : {e.Data}" : "");

        process.BeginOutputReadLine();
        process.BeginErrorReadLine();

        process!.WaitForExit();

        Assert.Equal(0, process.ExitCode);
    }

    private class GameInfo
    {
        public string GameFolder { get; set; } = null!;
        public string AsmFileToCompile { get; set; } = null!;
        public string NesOutputFile { get; set; } = null!;
    }
}
