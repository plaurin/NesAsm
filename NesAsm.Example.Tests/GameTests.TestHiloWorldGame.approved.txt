- Main diagnostics:
- SyntaxTrees Count: 2
- Output diagnostics:
- RunResult.GeneratedTrees Count: 1
- RunResults diagnostics:
- RunResult[0].Generator: Microsoft.CodeAnalysis.IncrementalGeneratorWrapper
- RunResult[0] diagnostics:
- RunResult[0].GeneratedSources Count: 1
- RunResult[0].Exception: 
- RunResult[0].GeneratedSources[0].HintName: Asm.HiloWorld.cs
- RunResult[0].GeneratedSources[0].SourceText:
/* Generator Asm code in file Asm.HiloWorld.s
; Auto generated code using the NesAsm project
.scope HiloWorld

  .proc main
    ; Read the PPUSTATUS register $2002
    lda $2002

    ; Store the address $3f01 in the PPUADDR $2006 (begining of the background palette 0)
    lda #$3F
    sta $2006
    lda #$00
    sta $2006

  @loop:

    lda palettes, x

    ; Write palette data to PPUDATA $2007
    sta $2007
    inx

    cpx #$20

    bne @loop

    rts
  .endproc

nmi:
  ldx #$00
  stx $2003

@spriteLoop:

  lda hilo_world_sprites, x
  sta $2004
  inx

  cpx #44
  bne @spriteLoop
  rti

.segment "CODE"

hilo_world_sprites:
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0

  .byte 30
  .byte 1
  .byte 0
  .byte 30
  .byte 30
  .byte 2
  .byte 0
  .byte 39

  .byte 30
  .byte 3
  .byte 0
  .byte 48
  .byte 30
  .byte 4
  .byte 0
  .byte 57

  .byte 40
  .byte 5
  .byte 0
  .byte 75
  .byte 40
  .byte 4
  .byte 0
  .byte 84

  .byte 40
  .byte 6
  .byte 0
  .byte 93
  .byte 40
  .byte 3
  .byte 0
  .byte 102

  .byte 40
  .byte 7
  .byte 0
  .byte 111

.segment "CODE"

palettes:
  .byte 15
  .byte 32
  .byte 33
  .byte 34
  .byte 15
  .byte 0
  .byte 0
  .byte 0

  .byte 15
  .byte 0
  .byte 0
  .byte 0
  .byte 15
  .byte 0
  .byte 0
  .byte 0

  .byte 15
  .byte 32
  .byte 39
  .byte 49
  .byte 15
  .byte 0
  .byte 0
  .byte 0

  .byte 15
  .byte 0
  .byte 0
  .byte 0
  .byte 15
  .byte 0
  .byte 0
  .byte 0


.segment "CHARS"

  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0

  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0

  .byte 195
  .byte 195
  .byte 195
  .byte 255
  .byte 255
  .byte 195
  .byte 195
  .byte 195

  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0

  .byte 255
  .byte 255
  .byte 24
  .byte 24
  .byte 24
  .byte 24
  .byte 255
  .byte 255

  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0

  .byte 192
  .byte 192
  .byte 192
  .byte 192
  .byte 192
  .byte 192
  .byte 255
  .byte 255

  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0

  .byte 24
  .byte 60
  .byte 102
  .byte 195
  .byte 195
  .byte 102
  .byte 60
  .byte 24

  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0

  .byte 195
  .byte 195
  .byte 195
  .byte 195
  .byte 102
  .byte 126
  .byte 102
  .byte 102

  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0

  .byte 248
  .byte 254
  .byte 195
  .byte 207
  .byte 248
  .byte 204
  .byte 198
  .byte 195

  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0

  .byte 248
  .byte 252
  .byte 198
  .byte 195
  .byte 195
  .byte 198
  .byte 252
  .byte 248

  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0
  .byte 0


.endscope

.include "wrapper-no-nmi.s"

*/
- RunResult[1].Generator: Microsoft.CodeAnalysis.IncrementalGeneratorWrapper
- RunResult[1] diagnostics:
- RunResult[1].GeneratedSources Count: 0
- RunResult[1].Exception: 
